ui/desktop/src/components/McpApps/useSandboxBridge.ts

```diff
@@ -0,0 +1,325 @@
+import { useRef, useEffect, useState, useCallback } from 'react';
+import { JsonRpcMessage, JsonRpcRequest, ToolInput, ToolResult } from './types';
+import {
+  fetchMcpAppProxyUrl,
+  createSandboxResourceReadyMessage,
+  createInitializeResponse,
+  createHostContextChangedNotification,
+  createToolInputNotification,
+  createToolResultNotification,
+  getCurrentTheme,
+  HostContext,
+} from './utils';
+
+interface SandboxBridgeOptions {
+  resourceHtml: string;
+  resourceCsp: Record<string, string[]> | null;
+  resourceUri: string;
+  toolInput?: ToolInput;
+  toolResult?: ToolResult;
+  appendMessage?: (value: string) => void;
+}
+
+interface SandboxBridgeResult {
+  iframeRef: React.RefObject<HTMLIFrameElement | null>;
+  proxyUrl: string | null;
+  iframeHeight: number;
+}
+
+export function useSandboxBridge(options: SandboxBridgeOptions): SandboxBridgeResult {
+  const { resourceHtml, resourceCsp, resourceUri, toolInput, toolResult, appendMessage } = options;
+
+  const iframeRef = useRef<HTMLIFrameElement | null>(null);
+  const pendingMessagesRef = useRef<JsonRpcMessage[]>([]);
+
+  const [proxyUrl, setProxyUrl] = useState<string | null>(null);
+  const [iframeHeight, setIframeHeight] = useState(200);
+  const [isSandboxReady, setIsSandboxReady] = useState(false);
+  const [isGuestInitialized, setIsGuestInitialized] = useState(false);
+
+  // Fetch proxy URL on mount
+  useEffect(() => {
+    fetchMcpAppProxyUrl().then(setProxyUrl);
+  }, []);
+
+  // Reset state when resource changes
+  useEffect(() => {
+    setIsSandboxReady(false);
+    setIsGuestInitialized(false);
+    pendingMessagesRef.current = [];
+  }, [resourceUri]);
+
+  const sendToSandbox = useCallback((message: JsonRpcMessage) => {
+    const iframe = iframeRef.current;
+    if (iframe?.contentWindow) {
+      iframe.contentWindow.postMessage(message, '*');
+    }
+  }, []);
+
+  const flushPendingMessages = useCallback(() => {
+    pendingMessagesRef.current.forEach((msg) => sendToSandbox(msg));
+    pendingMessagesRef.current = [];
+  }, [sendToSandbox]);
+
+  const handleMessage = useCallback(
```

Comment:
I think you are fighting the type system here. if you would just declare the types as something like

```
interface SandboxReadyNotification {
  jsonrpc: '2.0';
  method: 'ui/notifications/sandbox-ready';
}

interface InitializeRequest {
  jsonrpc: '2.0';
  id: string | number;
  method: 'ui/initialize';
}

...

type IncomingGuestMessage =
  | SandboxReadyNotification
  | InitializeRequest
  | InitializedNotification
  ...
```

you could then just switch on method and get this all for free:

```
switch (msg.method) {
    case 'ui/notifications/size-changed':
      setIframeHeight(Math.max(200, msg.params.height));
      break;
```